cpp/src/arrow/io/file.cc

```diff
@@ -526,13 +574,14 @@ Status MemoryMappedFile::Write(const void* data, int64_t nbytes) {
   if (!memory_map_->opened() || !memory_map_->writable()) {
     return Status::IOError("Unable to write");
   }
-  if (nbytes + memory_map_->position() > memory_map_->size()) {
-    return Status::Invalid("Cannot write past end of memory map");
-  }
   return WriteInternal(data, nbytes);
 }
 
 Status MemoryMappedFile::WriteInternal(const void* data, int64_t nbytes) {
+  if (nbytes + memory_map_->position() > memory_map_->capacity()) {
+    int64_t new_capacity = BitUtil::NextPower2(nbytes + memory_map_->position());
+    RETURN_NOT_OK(memory_map_->Resize(new_capacity));
```

Comment:
I'm somewhat ambivalent about this logic. If we're going to do this, we should also expose `Reserve` and `Resize` methods in `MemoryMappedFile`  so that the user has explicit control over how big the memory map gets

Reply:
Do you mean that power of 2 growth is too fast? I added `Resize` method so the user can truncate the file afterwards. Note that I hide the difference between `size` and `capacity` on the file level, so there is no `Reserve` method. 

May be we'd need an `ResizableFile` interface. Then we could think about adding resizing functionality to other types of files.